c/tskit/tables.c

```diff
@@ -7957,6 +7957,203 @@ tsk_table_collection_subset(
     return ret;
 }
 
+int TSK_WARN_UNUSED
+tsk_table_collection_union(tsk_table_collection_t *self, tsk_table_collection_t *other,
+    tsk_id_t *nodes, tsk_size_t num_nodes)
+{
+    int ret = 0;
+    tsk_id_t k, i, new_ind, new_pop, new_parent, new_child;
+    // tsk_id_t new_source, new_dest;
+    tsk_node_t node;
+    tsk_individual_t ind;
+    tsk_population_t pop;
+    tsk_edge_t edge;
+    tsk_mutation_t mut;
+    tsk_site_t site;
+    // tsk_migration_t mig;
+    tsk_id_t *node_map = NULL;
+    tsk_id_t *individual_map = NULL;
+    tsk_id_t *population_map = NULL;
+    tsk_id_t *site_map = NULL;
+    tsk_id_t *mutation_map = NULL;
+
+    // nodes relates id in other (index) to equivalent node in self or -1 if not present
+    if (num_nodes != other->nodes.num_rows) {
+        ret = TSK_ERR_BAD_MAPPING;
+        goto out;
+    }
+
+    // maps relates id in other (index) to new id in self
+    node_map = malloc(other->nodes.num_rows * sizeof(*node_map));
+    individual_map = malloc(other->individuals.num_rows * sizeof(*individual_map));
+    population_map = malloc(other->populations.num_rows * sizeof(*population_map));
+    site_map = malloc(other->sites.num_rows * sizeof(*site_map));
+    mutation_map = malloc(other->mutations.num_rows * sizeof(*mutation_map));
+    if (node_map == NULL || individual_map == NULL || population_map == NULL
+        || site_map == NULL || mutation_map == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    memset(node_map, 0xff, other->nodes.num_rows * sizeof(*node_map));
+    memset(individual_map, 0xff, other->individuals.num_rows * sizeof(*individual_map));
+    memset(population_map, 0xff, other->populations.num_rows * sizeof(*population_map));
+    memset(site_map, 0xff, other->sites.num_rows * sizeof(*site_map));
+    memset(mutation_map, 0xff, other->mutations.num_rows * sizeof(*mutation_map));
+
+    // nodes, individuals, populations
+    for (k = 0; k < (tsk_id_t) num_nodes; k++) {
+        if (nodes[k] != TSK_NULL) {
+            node_map[k] = nodes[k];
```

Comment:
should catch the error if `nodes[k] >= orig_num_self_nodes`, where `orig_self_num_nodes` was the nubmer of nodes in self at the start